plugins/redfish/fu-ipmi-device.c

```diff
@@ -0,0 +1,573 @@
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <fcntl.h>
+#include <glib/gstdio.h>
+#include <linux/ipmi.h>
+#include <linux/ipmi_msgdefs.h>
+#include <poll.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "fu-ipmi-device.h"
+
+#define FU_IPMI_DEVICE_TIMEOUT 1500 /* ms */
+
+/* not defined in linux/ipmi_msgdefs.h */
+#define IPMI_SET_USER_ACCESS   0x43
+#define IPMI_SET_USER_NAME     0x45
+#define IPMI_GET_USER_NAME     0x46
+#define IPMI_SET_USER_PASSWORD 0x47
+
+#define IPMI_PASSWORD_DISABLE_USER  0x00
+#define IPMI_PASSWORD_ENABLE_USER   0x01
+#define IPMI_PASSWORD_SET_PASSWORD  0x02
+#define IPMI_PASSWORD_TEST_PASSWORD 0x03
+
+struct _FuIpmiDevice {
+	FuUdevDevice parent_instance;
+	glong seq;
+	guint8 device_id;
+	guint8 device_rev;
+	guint8 version_ipmi;
+};
+
+G_DEFINE_TYPE(FuIpmiDevice, fu_ipmi_device, FU_TYPE_UDEV_DEVICE)
+
+static void
+fu_ipmi_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuIpmiDevice *self = FU_IPMI_DEVICE(device);
+	fu_common_string_append_kx(str, idt, "DeviceId", self->device_id);
+	fu_common_string_append_kx(str, idt, "DeviceRev", self->device_rev);
+	fu_common_string_append_kx(str, idt, "VersionIpmi", self->version_ipmi);
+}
+
+static gboolean
+fu_ipmi_device_send(FuIpmiDevice *self,
+		    guint8 netfn,
+		    guint8 cmd,
+		    const guint8 *buf,
+		    gsize bufsz,
+		    GError **error)
+{
+	g_autofree guint8 *buf2 = fu_memdup_safe(buf, bufsz, NULL);
+	struct ipmi_system_interface_addr addr = {.addr_type = IPMI_SYSTEM_INTERFACE_ADDR_TYPE,
+						  .channel = IPMI_BMC_CHANNEL};
+	struct ipmi_req req = {
+	    .addr = (guint8 *)&addr,
+	    .addr_len = sizeof(addr),
+	    .msgid = self->seq++,
+	    .msg.data = buf2,
+	    .msg.data_len = (guint16)bufsz,
+	    .msg.netfn = netfn,
+	    .msg.cmd = cmd,
+	};
+	if (g_getenv("FWUPD_REDFISH_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "ipmi-send", buf2, bufsz);
+	return fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				    IPMICTL_SEND_COMMAND,
+				    (guint8 *)&req,
+				    NULL,
+				    error);
+}
+
+static gboolean
+fu_ipmi_device_recv(FuIpmiDevice *self,
+		    guint8 *netfn,
+		    guint8 *cmd,
+		    glong *seq,
+		    guint8 *buf,
+		    gsize bufsz,
+		    gsize *len, /* optional, out */
+		    GError **error)
+{
+	struct ipmi_addr addr = {0};
+	struct ipmi_recv recv = {
+	    .addr = (guint8 *)&addr,
+	    .addr_len = sizeof(addr),
+	    .msg.data = buf,
+	    .msg.data_len = bufsz,
+	};
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  IPMICTL_RECEIVE_MSG_TRUNC,
+				  (guint8 *)&recv,
+				  NULL,
+				  error))
+		return FALSE;
+	if (g_getenv("FWUPD_REDFISH_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "ipmi-recv", buf, *len);
+	if (netfn != NULL)
+		*netfn = recv.msg.netfn;
+	if (cmd != NULL)
+		*cmd = recv.msg.cmd;
+	if (seq != NULL)
+		*seq = recv.msgid;
+	if (len != NULL)
+		*len = (gsize)recv.msg.data_len;
```

Comment:
should any of these set an error?

Reply:
I don't _think_ so. `fu_ipmi_device_recv` just gets the response, no matter the contents.

Reply:
Okay.